sdks/python/apache_beam/runners/interactive/interactive_beam.py

```diff
@@ -82,7 +89,100 @@ def run_pipeline(self):
+
+  The function is always blocking. If used within a notebook, the data
+  visualized might be dynamically updated before the function returns as more
+  and more data could getting processed and emitted when the pipeline fragment
+  is being executed. If used within an ipython shell, there will be no dynamic
+  plotting but a static plotting in the end of pipeline fragment execution.
+
+  The PCollections given must belong to the same pipeline and be watched by
+  Interactive Beam (PCollections defined in __main__ are automatically watched).
+
+    For example::
+
+      p = beam.Pipeline(InteractiveRunner())
+      init = p | 'Init' >> beam.Create(range(1000))
+      square = init | 'Square' >> beam.Map(lambda x: x * x)
+      cube = init | 'Cube' >> beam.Map(lambda x: x ** 3)
+
+      # Below builds a pipeline fragment from the defined pipeline `p` that
+      # contains only applied transforms of `Init` and `Square`. Then the
+      # interactive runner runs the pipeline fragment implicitly to compute data
+      # represented by PCollection `square` and visualizes it.
+      show(square)
+
+      # This is equivalent to `show(square)` because `square` depends on `init`
+      # and `init` is included in the pipeline fragment and computed anyway.
+      show(init, square)
+
+      # Below is similar to running `p.run()`. It computes data for both
+      # PCollection `square` and PCollection `cube`, then visualizes them.
+      show(square, cube)
+  """
+  assert len(pcolls) > 0, (
+      'Need at least 1 PCollection to show data visualization.')
+  for pcoll in pcolls:
+    assert isinstance(pcoll, beam.pvalue.PCollection), (
+        '{} is not an apache_beam.pvalue.PCollection.'.format(pcoll))
+  user_pipeline = pcolls[0].pipeline
+  for pcoll in pcolls:
+    assert pcoll.pipeline is user_pipeline, (
+        '{} belongs to a different user-defined pipeline ({}) than that of'
+        ' other PCollections ({}).'.format(
+            pcoll, pcoll.pipeline, user_pipeline))
+  runner = user_pipeline.runner
+  if isinstance(runner, ir.InteractiveRunner):
+    runner = runner._underlying_runner
+
+  # Make sure that all PCollections to be shown are watched. If a PCollection
+  # has not been watched, make up a variable name for that PCollection and watch
+  # it. No validation is needed here because the watch logic can handle
+  # arbitrary variables.
+  watched_pcollections = set()
+  for watching in ie.current_env().watching():
+    for key, val in watching:
+      if hasattr(val, '__class__') and isinstance(val, beam.pvalue.PCollection):
+        watched_pcollections.add(val)
+  for pcoll in pcolls:
+    if pcoll not in watched_pcollections:
+      watch({re.sub(r'[\[\]\(\)]', '_', str(pcoll)): pcoll})
+
+  # Attempt to run background caching job since we have the reference to the
+  # user-defined pipeline.
+  bcj.attempt_to_run_background_caching_job(runner, user_pipeline)
```

Comment:
You renamed `pin` to `build_pipeline_instrument` in another place. Do you want to switch these to `background_caching_job` if you are trying to be more explicit in general?

Reply:
This is to discern instances from modules.
For example, given a `background_caching_job`, it feels more like an instance of `BackgroundCachingJob`.
For `pipeline_instrument`, when importing the module, we normally rename it to `instr` or `inst`.
Because `pipeline_instrument = instr.build_pipeline_instrument(...)`.
The abbreviation is to avoid name conflicts.